File: problems/java/ParkingLotSystem/src/model/ParkingTicket.java
package model;

import java.util.UUID;

/**
 * Links a parked vehicle to its spot and entry/exit times. Used for fee calculation and to
 * release the spot on exit (after successful payment).
 */
public class ParkingTicket {

    private final String ticketId;
    private final Vehicle parkedVehicle;
    private final ParkingSpot assignedSpot;
    private final long startTimeStamp;
    private long endTimeStamp;

    public ParkingTicket(Vehicle vehicle, ParkingSpot spot) {
        this.parkedVehicle = vehicle;
        this.assignedSpot = spot;
        this.ticketId = UUID.randomUUID().toString();
        this.startTimeStamp = System.currentTimeMillis();
    }

    public String getTicketId() {
        return ticketId;
    }

    public ParkingSpot getAssignedSpot() {
        return assignedSpot;
    }

    /** Duration from start until end (or now if ticket not yet closed). */
    public double getDurationInHours() {
        long endTime = (endTimeStamp == 0) ? System.currentTimeMillis() : endTimeStamp;
        return (double) (endTime - startTimeStamp) / (1000 * 60 * 60);
    }

    /**
     * Sets the exit time. Called at exit before fee calculation.
     */
    public void setEndTimeStamp() {
        endTimeStamp = System.currentTimeMillis();
    }
}

File: problems/python/ParkingLotSystem/src/model/parking_ticket.py
"""Ticket linking vehicle to spot and timestamps."""
from __future__ import annotations
import uuid
import time
from model.vehicle import Vehicle


class ParkingTicket:
    def __init__(self, vehicle: Vehicle, spot: "ParkingSpot"):
        self._parked_vehicle = vehicle
        self._assigned_spot = spot
        self._ticket_id = str(uuid.uuid4())
        self._start_time_stamp = time.time() * 1000
        self._end_time_stamp: float = 0

    def get_ticket_id(self) -> str:
        return self._ticket_id

    def get_assigned_spot(self) -> "ParkingSpot":
        return self._assigned_spot

    def get_duration_in_hours(self) -> float:
        end_time = self._end_time_stamp if self._end_time_stamp else time.time() * 1000
        return (end_time - self._start_time_stamp) / (1000 * 60 * 60)

    def set_end_time_stamp(self) -> None:
        self._end_time_stamp = time.time() * 1000
